File: ben/haskell/Q11-20-lists-cont/p18.hs
-- With prelude 
slice :: [a] -> Int -> Int -> [a] 
slice list begin end = take (end - begin + 1) . drop (begin - 1) $ list

-- With list recursion
slice' :: [a] -> Int -> Int -> [a] 
slice' list 1 0 = []
slice' (x:xs) 1 nb = x : slice' xs 1 (nb-1)
slice' (_:xs) n1 n2 = slice' xs (n1-1) (n2-1) 

-- With list comprehension 
slice'' :: [a] -> Int -> Int -> [a] 
slice'' list n1 n2 = [list!!i | i <- [0..(length list - 1)],
                                i >= (n1-1), 
                                i < n2]

File: ben/prolog/P01-P28-lists/p18.pl
% Extract a slice from a list.

slice([X|_], _, 1, [X]).
slice([X|Xs], 1, N0, [X|R]) :-
    N0 > 1,
    N1 is N0 - 1, 
    slice(Xs, 1, N1, R).
slice([_|Xs], N0, N1, R) :-
    N0 > 1, 
    N1 > 1,
    N2 is N0 - 1,
    N3 is N1 - 1,
    slice(Xs, N2, N3, R).
